### PR TITLE
Make ant work with javac8 and JDK7

### PR DESCRIPTION
**Pull request by cushon**
This generalizes some classloading infrastructure that could be shared
with maven.

### CODE REVIEW

**Comment by cushon** [comment]
This also fixes the less interesting part of #287

File: core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+package com.google.errorprone.internal;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.net.URLClassLoader;
+import java.util.Set;
+
+/**
+ * A non-delegating {@link java.net.URLClassLoader} that searches it's own resource path _before_
```
Comment (eaftan) [inline core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java]: nit: it's -> its

File: core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * @author cushon@google.com
+ */
+public class NonDelegatingClassLoader extends URLClassLoader {
+  private final ClassLoader original;
+  private final ImmutableSet<String> whiteList;
+
+  public static NonDelegatingClassLoader create(Set<String> whiteList) {
+    ClassLoader contextLoader = Thread.currentThread().getContextClassLoader();
+    if (!(contextLoader instanceof URLClassLoader)) {
```
Comment (eaftan) [inline core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java]: Can we always be sure the default classloader is a URLClassLoader?

File: core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java (hunk 1 above)
Comment (cushon) [inline core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java]: Fixed.

Heh, github markdown recognized `_before_` and `<i>before<i>` as _before_.

File: core/src/main/java/com/google/errorprone/internal/SandboxedClassLoaderRunner.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * Executes a callback inside a special classloader that searches its own resource path _before_
+ * delegating to the runtime classpath. This allows error-prone's javac8 to be run on JDK7, and is
+ * morally equivalent to invoking java with {@code -Xbootclasspath/p:<path to javac.jar>}.
+ *
+ * @author cushon@google.com
+ */
+public class SandboxedClassLoaderRunner {
```
Comment (eaftan) [inline core/src/main/java/com/google/errorprone/internal/SandboxedClassLoaderRunner.java]: nit: maybe rename to something like NonDelegatingClassLoaderRunner?  "Sandboxed" implies that this is security related.

File: core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java (hunk 2 above)
Comment (cushon) [inline core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java]: This assumption is unfortunate, but it holds for all known plugins and I don't know any reasonable alternatives.

Comment (eaftan) [inline core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java]: OK, fine with me

**Comment by eaftan** [comment]
LGTM aside from the nits.  Go ahead and merge when you're done.

File: core/src/main/java/com/google/errorprone/internal/SandboxedClassLoaderRunner.java (hunk 3 above)
Comment (cushon) [inline core/src/main/java/com/google/errorprone/internal/SandboxedClassLoaderRunner.java]: Done

**Pull request merged**